README.md
# `OTP`: HMAC-Based and Time-Based One-Time Passwords (HOTP & TOTP)  [![Hackage](https://img.shields.io/hackage/v/OTP.svg)](https://hackage.haskell.org/package/OTP) [![Build Status](https://travis-ci.org/haskell-hvr/OTP.svg)](https://travis-ci.org/haskell-hvr/OTP)

**Please refer to the [package description](https://hackage.haskell.org/package/OTP#description) for an overview of `OTP`.**

## Usage examples

### Generating one-time passwords

If you need to generate HOTP password described in RFC4226, then use

```haskell
>>> hotp SHA1 "1234" 100 6
317569

>>> hotp SHA512 "1234" 100 6
134131
```

Or

```haskell
>>> totp SHA1 "1234" (read "2010-10-10 00:01:00 UTC") 30 8
43388892
```

to generate TOTP password described in RFC6238.

### Checking one-time passwords

```haskell
hotpCheck :: HashAlgorithm      -- ^ Hashing algorithm
          -> Secret             -- ^ Shared secret
          -> (Word8, Word8)     -- ^ how much counters to take lower and higher than ideal
          -> Word64             -- ^ ideal (expected) counter value
          -> Word8              -- ^ Number of digits in password
          -> Word32             -- ^ Password entered by user
          -> Bool               -- ^ True if password acceptable
```

```haskell
>>> hotpCheck SHA1 "1234" (0,0) 10 6 50897
True

>>> hotpCheck SHA1 "1234" (0,0) 9 6 50897
False

>>> hotpCheck SHA1 "1234" (0,1) 9 6 50897
True
```

Here almost the same aguments as for `hotp` function, but there is
also `(0, 0)` tuple. This tuple describes range of counters to check
in case of desynchronisation of counters between client and
server. I.e. if you specify `(1, 1)` and ideal counter will be `10`
then function will check passwords for `[9, 10, 11]` list of
counters.

Here is the same for TOTP:

```haskell
>>> totpCheck SHA1 "1234" (0, 0) (read "2010-10-10 00:00:00 UTC") 30 6 778374
True

>>> totpCheck SHA1 "1234" (0, 0) (read "2010-10-10 00:00:30 UTC") 30 6 778374
False

>>> totpCheck SHA1 "1234" (1, 0) (read "2010-10-10 00:00:30 UTC") 30 6 778374
True
```


CHANGELOG.md
## 0.1.0.0

* Major API rewrite ([#3](https://github.com/hvr/OTP/pull/3))
* Add support for SHA256 and SHA512 besides SHA-1 ([#1](https://github.com/hvr/OTP/pull/1))
* Provide support via cabal flag for using pure Haskell [SHA](http://hackage.haskell.org/package/SHA) implementation

## 0.0.0.1

Initial version of `OTP`


Test/opt.hs
module Main where

import Codec.Utils
import Data.OTP
import Data.Time
import System.Exit (exitFailure)

secret_otp :: [Octet]
secret_otp = [49,50,51,52,53,54,55,56,57,48,49,50,51,52,53,54,55,56,57,48]

hotp_t = zipWith (\x y -> (hotp secret_otp x 6) == y) [0..9]
        [755224, 287082, 359152, 969429, 338314, 254676, 287922, 162583, 399871, 520489]

totp_t = zipWith (\x y -> (totp secret_otp x 8 30) == y) (map (read) ["1970-01-01 00:00:59 UTC",
        "2005-03-18 01:58:29 UTC", "2005-03-18 01:58:31 UTC", "2009-02-13 23:31:30 UTC",
        "2033-05-18 03:33:20 UTC"])
        [94287082, 07081804, 14050471, 89005924, 69279037]
tests = hotp_t ++ totp_t

main =  if (and tests) then putStrLn "Tests succeful!"
        else do
                putStrLn "Tests falls!"
                exitFailure


test/opt.hs
module Main where

import Control.Arrow
import Crypto.Hash
import Data.ByteString  (ByteString)
import Data.OTP
import Data.Time
import Data.Word
import System.Exit      (exitFailure)
import Test.Tasty
import Test.Tasty.HUnit

import qualified Data.ByteString as BS
import qualified Data.ByteString.Char8 as BC

hotpSecret :: ByteString
hotpSecret = "12345678901234567890"

testHotp :: Word64 -> Word32 -> TestTree
testHotp key result = testCase (show result) $ do
    let h = hotp SHA1 hotpSecret key 6
    result @=? h

hotpResults :: [Word32]
hotpResults =
    [ 755224, 287082, 359152
    , 969429, 338314, 254676
    , 287922, 162583, 399871
    , 520489
    ]


data SomeAlg = forall a. (HashAlgorithm a, Show a) => SomeAlg { getAlg :: a }

instance Show SomeAlg where
    show (SomeAlg a) = show a


testTotp :: (ByteString, UTCTime, SomeAlg, Word32) -> TestTree
testTotp (secr, key, alg', result) =
    testCase (show alg' ++ " => " ++ show result) $ case alg' of
        SomeAlg alg -> do
            let t = totp alg secr key 30 8
            result @=? t

sha1Secr   = BC.pack $ take 20 $ cycle "12345678901234567890"
sha256Secr = BC.pack $ take 32 $ cycle "12345678901234567890"
sha512Secr = BC.pack $ take 64 $ cycle "12345678901234567890"

totpData :: [(ByteString, UTCTime, SomeAlg, Word32)]
totpData =
    [ (sha1Secr,   read "1970-01-01 00:00:59 UTC", SomeAlg SHA1,   94287082)
    , (sha256Secr, read "1970-01-01 00:00:59 UTC", SomeAlg SHA256, 46119246)
    , (sha512Secr, read "1970-01-01 00:00:59 UTC", SomeAlg SHA512, 90693936)
    , (sha1Secr,   read "2005-03-18 01:58:29 UTC", SomeAlg SHA1,   07081804)
    , (sha256Secr, read "2005-03-18 01:58:29 UTC", SomeAlg SHA256, 68084774)
    , (sha512Secr, read "2005-03-18 01:58:29 UTC", SomeAlg SHA512, 25091201)
    , (sha1Secr,   read "2005-03-18 01:58:31 UTC", SomeAlg SHA1,   14050471)
    , (sha256Secr, read "2005-03-18 01:58:31 UTC", SomeAlg SHA256, 67062674)
    , (sha512Secr, read "2005-03-18 01:58:31 UTC", SomeAlg SHA512, 99943326)
    , (sha1Secr,   read "2009-02-13 23:31:30 UTC", SomeAlg SHA1,   89005924)
    , (sha256Secr, read "2009-02-13 23:31:30 UTC", SomeAlg SHA256, 91819424)
    , (sha512Secr, read "2009-02-13 23:31:30 UTC", SomeAlg SHA512, 93441116)
    , (sha1Secr,   read "2033-05-18 03:33:20 UTC", SomeAlg SHA1,   69279037)
    , (sha256Secr, read "2033-05-18 03:33:20 UTC", SomeAlg SHA256, 90698825)
    , (sha512Secr, read "2033-05-18 03:33:20 UTC", SomeAlg SHA512, 38618901)
    , (sha1Secr,   read "2603-10-11 11:33:20 UTC", SomeAlg SHA1,   65353130)
    , (sha256Secr, read "2603-10-11 11:33:20 UTC", SomeAlg SHA256, 77737706)
    , (sha512Secr, read "2603-10-11 11:33:20 UTC", SomeAlg SHA512, 47863826)
    ]


main :: IO ()
main = defaultMain $ testGroup "unit tests"
    [ testGroup "hotp" $ map (uncurry testHotp) $ zip [0..] hotpResults
    , testGroup "totp" $ map testTotp totpData
    ]


test/Test.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import           Data.ByteString       (ByteString)
import qualified Data.ByteString.Char8 as BC
import           Data.Time
import           Data.Word
import           Test.Tasty
import           Test.Tasty.HUnit

-- IUT
import           Data.OTP


hotpSecret :: ByteString
hotpSecret = "12345678901234567890"

testHotp :: Word64 -> Word32 -> TestTree
testHotp key result = testCase (show result) $ do
    let h = hotp SHA1 hotpSecret key 6
    result @=? h

hotpResults :: [Word32]
hotpResults =
    [ 755224, 287082, 359152
    , 969429, 338314, 254676
    , 287922, 162583, 399871
    , 520489
    ]


testTotp :: (ByteString, UTCTime, HashAlgorithm, Word32) -> TestTree
testTotp (secr, key, alg, result) =
    testCase (show alg ++ " => " ++ show result) $ do
            let t = totp alg secr key 30 8
            result @=? t

sha1Secr :: ByteString
sha1Secr   = BC.pack $ take 20 $ cycle "12345678901234567890"

sha256Secr :: ByteString
sha256Secr = BC.pack $ take 32 $ cycle "12345678901234567890"

sha512Secr :: ByteString
sha512Secr = BC.pack $ take 64 $ cycle "12345678901234567890"

totpData :: [(ByteString, UTCTime, HashAlgorithm, Word32)]
totpData =
    [ (sha1Secr,   read "1970-01-01 00:00:59 UTC", SHA1,   94287082)
    , (sha256Secr, read "1970-01-01 00:00:59 UTC", SHA256, 46119246)
    , (sha512Secr, read "1970-01-01 00:00:59 UTC", SHA512, 90693936)
    , (sha1Secr,   read "2005-03-18 01:58:29 UTC", SHA1,   07081804)
    , (sha256Secr, read "2005-03-18 01:58:29 UTC", SHA256, 68084774)
    , (sha512Secr, read "2005-03-18 01:58:29 UTC", SHA512, 25091201)
    , (sha1Secr,   read "2005-03-18 01:58:31 UTC", SHA1,   14050471)
    , (sha256Secr, read "2005-03-18 01:58:31 UTC", SHA256, 67062674)
    , (sha512Secr, read "2005-03-18 01:58:31 UTC", SHA512, 99943326)
    , (sha1Secr,   read "2009-02-13 23:31:30 UTC", SHA1,   89005924)
    , (sha256Secr, read "2009-02-13 23:31:30 UTC", SHA256, 91819424)
    , (sha512Secr, read "2009-02-13 23:31:30 UTC", SHA512, 93441116)
    , (sha1Secr,   read "2033-05-18 03:33:20 UTC", SHA1,   69279037)
    , (sha256Secr, read "2033-05-18 03:33:20 UTC", SHA256, 90698825)
    , (sha512Secr, read "2033-05-18 03:33:20 UTC", SHA512, 38618901)
    , (sha1Secr,   read "2603-10-11 11:33:20 UTC", SHA1,   65353130)
    , (sha256Secr, read "2603-10-11 11:33:20 UTC", SHA256, 77737706)
    , (sha512Secr, read "2603-10-11 11:33:20 UTC", SHA512, 47863826)
    ]


main :: IO ()
main = defaultMain $ testGroup "test vectors"
    [ testGroup "hotp" $ map (uncurry testHotp) $ zip [0..] hotpResults
    , testGroup "totp" $ map testTotp totpData
    ]


src/HashImpl.hs
{-# LANGUAGE CPP #-}

-- | SPDX-License-Identifier: MIT
--
-- @since 0.1.0.0
module HashImpl where

#if defined(MIN_VERSION_SHA)
import qualified Data.ByteString.Lazy as BS.L
import qualified Data.Digest.Pure.SHA as SHA
#else
import qualified Crypto.Hash.SHA1     as SHA1
import qualified Crypto.Hash.SHA256   as SHA256
import qualified Crypto.Hash.SHA512   as SHA512
#endif

import qualified Data.ByteString      as BS


-- | Shared secret encoded as raw octets
type Secret = BS.ByteString

-- | Hash algorithm used for HOTP\/TOTP computations
data HashAlgorithm = SHA1
                   | SHA256
                   | SHA512
                   deriving (Eq,Show)

hmac :: HashAlgorithm -> Secret -> BS.ByteString -> BS.ByteString
hmac alg key msg = case alg of
#if defined(MIN_VERSION_SHA)
    SHA1   -> BS.L.toStrict (SHA.bytestringDigest (SHA.hmacSha1   (BS.L.fromStrict key) (BS.L.fromStrict msg)))
    SHA256 -> BS.L.toStrict (SHA.bytestringDigest (SHA.hmacSha256 (BS.L.fromStrict key) (BS.L.fromStrict msg)))
    SHA512 -> BS.L.toStrict (SHA.bytestringDigest (SHA.hmacSha512 (BS.L.fromStrict key) (BS.L.fromStrict msg)))
#else
    SHA1   -> SHA1.hmac   key msg
    SHA256 -> SHA256.hmac key msg
    SHA512 -> SHA512.hmac key msg
#endif


src/Data/OTP.hs
-- | SPDX-License-Identifier: MIT
--
-- Implements the /HMAC-Based One-Time Password Algorithm/ (HOTP) as
-- defined in [RFC 4226](https://tools.ietf.org/html/rfc4226)
-- and the /Time-Based One-Time Password Algorithm/ (TOTP) as defined
-- in [RFC 6238](https://tools.ietf.org/html/rfc6238).
--
-- Many operations in this module take or return a 'Word32' OTP value
-- (whose most significant bit is always 0) which is truncated modulo
-- @10^digits@ according to the 'Word8' /digits/
-- parameter. Consequently, passing a value above 10 won't produce
-- more than 10 digits and will effectively return the raw
-- non-truncated 31-bit OTP value.
--
-- @since 0.1.0.0
module Data.OTP
       ( -- * HOTP
         hotp
       , hotpCheck
         -- * TOTP
       , totp
       , totpCheck
         -- * Auxiliary
       , totpCounter
       , counterRange
       , totpCounterRange

       , HashAlgorithm(..)
       , Secret
       ) where

import           Data.Bits
import qualified Data.ByteString       as BS
import           Data.Time
import           Data.Time.Clock.POSIX
import           Data.Word

import           HashImpl

{- | Compute /HMAC-Based One-Time Password/ using secret key and counter value.

>>> hotp SHA1 "1234" 100 6
317569

>>> hotp SHA512 "1234" 100 6
134131

>>> hotp SHA512 "1234" 100 8
55134131

-}

hotp
  :: HashAlgorithm           -- ^ Hashing algorithm
  -> Secret                  -- ^ Shared secret
  -> Word64                  -- ^ Counter value
  -> Word8                   -- ^ Number of base10 digits in HOTP value
  -> Word32                  -- ^ HOTP value
hotp alg key cnt digits
  | digits >= 10 = snum
  | otherwise    = snum `rem` (10 ^ digits)
  where
    -- C
    msg = bsFromW64 cnt

    -- Snum  = StToNum(Sbits)
    -- Sbits = DT(HS)
    -- HS    = HMAC(K,C)
    snum = trunc $ hmac alg key msg

    -- DT(HS)
    trunc :: BS.ByteString -> Word32
    trunc b = case bsToW32 rb of
                Left e    -> error e
                Right res -> res .&. (0x80000000 - 1) -- reset highest bit
      where
        offset = BS.last b .&. 15 -- take low 4 bits of last byte
        rb = BS.take 4 $ BS.drop (fromIntegral offset) b -- resulting 4 byte value

    -- StToNum(Sbits)
    bsToW32 :: BS.ByteString -> Either String Word32
    bsToW32 bs = case BS.unpack bs of
                   [ b0, b1, b2, b3 ] -> Right $! (((((fI b0 `shiftL` 8) .|. fI b1) `shiftL` 8) .|. fI b2) `shiftL` 8) .|. fI b3
                   _                  -> Left "bsToW32: the impossible happened"
      where
        fI = fromIntegral

    bsFromW64 :: Word64 -> BS.ByteString
    bsFromW64 w = BS.pack [ b j | j <- [ 7, 6 .. 0 ] ]
      where
        b j = fromIntegral (w `shiftR` (j*8))

{- | Check presented password against a valid range.

>>> hotp SHA1 "1234" 10 6
50897

>>> hotpCheck SHA1 "1234" (0,0) 10 6 50897
True

>>> hotpCheck SHA1 "1234" (0,0) 9 6 50897
False

>>> hotpCheck SHA1 "1234" (0,1) 9 6 50897
True

>>> hotpCheck SHA1 "1234" (1,0) 11 6 50897
True

>>> hotpCheck SHA1 "1234" (2,2) 8 6 50897
True

>>> hotpCheck SHA1 "1234" (2,2) 7 6 50897
False

>>> hotpCheck SHA1 "1234" (2,2) 12 6 50897
True

>>> hotpCheck SHA1 "1234" (2,2) 13 6 50897
False

-}

hotpCheck
  :: HashAlgorithm      -- ^ Hash algorithm to use
  -> Secret             -- ^ Shared secret
  -> (Word8, Word8)     -- ^ Valid counter range, before and after ideal
  -> Word64             -- ^ Ideal (expected) counter value
  -> Word8              -- ^ Number of base10 digits in a password
  -> Word32             -- ^ Password (i.e. HOTP value) entered by user
  -> Bool               -- ^ True if password is valid
hotpCheck alg secr rng cnt len pass =
    let counters = counterRange rng cnt
        passwds = map (\c -> hotp alg secr c len) counters
    in any (pass ==) passwds

{- | Compute a /Time-Based One-Time Password/ using secret key and time.

>>> totp SHA1 "1234" (read "2010-10-10 00:01:00 UTC") 30 6
388892

>>> totp SHA1 "1234" (read "2010-10-10 00:01:00 UTC") 30 8
43388892

>>> totp SHA1 "1234" (read "2010-10-10 00:01:15 UTC") 30 8
43388892

>>> totp SHA1 "1234" (read "2010-10-10 00:01:31 UTC") 30 8
39110359

-}

totp
  :: HashAlgorithm -- ^ Hash algorithm to use
  -> Secret    -- ^ Shared secret
  -> UTCTime   -- ^ Time of TOTP
  -> Word64    -- ^ Time range in seconds
  -> Word8     -- ^ Number of base10 digits in TOTP value
  -> Word32    -- ^ TOTP value
totp alg secr time period len =
    hotp alg secr (totpCounter time period) len

{- | Check presented password against time periods.

>>> totp SHA1 "1234" (read "2010-10-10 00:00:00 UTC") 30 6
778374

>>> totpCheck SHA1 "1234" (0, 0) (read "2010-10-10 00:00:00 UTC") 30 6 778374
True

>>> totpCheck SHA1 "1234" (0, 0) (read "2010-10-10 00:00:30 UTC") 30 6 778374
False

>>> totpCheck SHA1 "1234" (1, 0) (read "2010-10-10 00:00:30 UTC") 30 6 778374
True

>>> totpCheck SHA1 "1234" (1, 0) (read "2010-10-10 00:01:00 UTC") 30 6 778374
False

>>> totpCheck SHA1 "1234" (2, 0) (read "2010-10-10 00:01:00 UTC") 30 6 778374
True
-}

totpCheck
  :: HashAlgorithm      -- ^ Hash algorithm to use
  -> Secret             -- ^ Shared secret
  -> (Word8, Word8)     -- ^ Valid counter range, before and after ideal
  -> UTCTime            -- ^ Time of TOTP
  -> Word64             -- ^ Time range in seconds
  -> Word8              -- ^ Number of base10 digits in a password
  -> Word32             -- ^ Password given by user
  -> Bool               -- ^ True if password is valid
totpCheck alg secr rng time period len pass =
    let counters = totpCounterRange rng time period
        passwds = map (\c -> hotp alg secr c len) counters
    in any (pass ==) passwds


{- | Calculate HOTP counter using time. Starting time (T0
according to RFC6238) is 0 (begining of UNIX epoch)

>>> totpCounter (read "2010-10-10 00:00:00 UTC") 30
42888960

>>> totpCounter (read "2010-10-10 00:00:30 UTC") 30
42888961

>>> totpCounter (read "2010-10-10 00:01:00 UTC") 30
42888962

-}

totpCounter
  :: UTCTime     -- ^ Time of totp
  -> Word64      -- ^ Time range in seconds
  -> Word64      -- ^ Resulting counter
totpCounter time period =
    let timePOSIX = floor $ utcTimeToPOSIXSeconds time
    in timePOSIX `div` period

{- | Make a sequence of acceptable counters, protected from
arithmetic overflow.

>>> counterRange (0, 0) 9000
[9000]

>>> counterRange (1, 0) 9000
[8999,9000]

>>> length $ counterRange (5000, 0) 9000
501

>>> length $ counterRange (5000, 5000) 9000
1000

>>> counterRange (2, 2) maxBound
[18446744073709551613,18446744073709551614,18446744073709551615]

>>> counterRange (2, 2) minBound
[0,1,2]

>>> counterRange (2, 2) (maxBound `div` 2)
[9223372036854775805,9223372036854775806,9223372036854775807,9223372036854775808,9223372036854775809]

>>> counterRange (5, 5) 9000
[8995,8996,8997,8998,8999,9000,9001,9002,9003,9004,9005]

RFC recommends avoiding excessively large values for counter ranges.
-}

counterRange
  :: (Word8, Word8)   -- ^ Number of counters before and after ideal
  -> Word64           -- ^ Ideal counter value
  -> [Word64]
counterRange (tolow, tohigh) ideal = [l..h]
  where
    l' = ideal - fromIntegral tolow
    l | l' <= ideal = l'
      | otherwise  = 0

    h' = ideal + fromIntegral tohigh
    h | ideal <= h' = h'
      | otherwise   = maxBound

{- | Make a sequence of acceptable periods.

>>> totpCounterRange (0, 0) (read "2010-10-10 00:01:00 UTC") 30
[42888962]

>>> totpCounterRange (2, 0) (read "2010-10-10 00:01:00 UTC") 30
[42888960,42888961,42888962]

>>> totpCounterRange (0, 2) (read "2010-10-10 00:01:00 UTC") 30
[42888962,42888963,42888964]

>>> totpCounterRange (2, 2) (read "2010-10-10 00:01:00 UTC") 30
[42888960,42888961,42888962,42888963,42888964]

-}

totpCounterRange :: (Word8, Word8)
                 -> UTCTime
                 -> Word64
                 -> [Word64]
totpCounterRange rng time period =
    counterRange rng $ totpCounter time period
